pallets/capacity/src/lib.rs

```diff
@@ -0,0 +1,658 @@
+	pub enum Event<T: Config> {
+		/// Tokens have been staked to the Frequency network.
+		Staked {
+			/// The token account that staked tokens to the network.
+			account: T::AccountId,
+			/// The MSA that a token account targeted to receive Capacity based on this staking amount.
+			target: MessageSourceId,
+			/// An amount that was staked.
+			amount: BalanceOf<T>,
+			/// The Capacity amount issued to the target as a result of the stake.
+			capacity: BalanceOf<T>,
+		},
+		/// Unsstaked token that has thawed was unlocked for the given account
+		StakeWithdrawn {
+			/// the account that withdrew its stake
+			account: T::AccountId,
+			/// the total amount withdrawn, i.e. put back into free balance.
+			amount: BalanceOf<T>,
+		},
+		/// A token account has unstaked the Frequency network.
+		UnStaked {
+			/// The token account that staked tokens to the network.
+			account: T::AccountId,
+			/// The MSA that a token account targeted to receive Capacity to unstake from.
+			target: MessageSourceId,
+			/// An amount that was unstaked.
+			amount: BalanceOf<T>,
+			/// The Capacity amount that was reduced from a target.
+			capacity: BalanceOf<T>,
+		},
+		/// The Capacity epoch length was changed.
+		EpochLengthUpdated {
+			/// The new length of an epoch in blocks.
+			blocks: T::BlockNumber,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Staker attempted to stake to an invalid staking target.
+		InvalidTarget,
+		/// Staker has insufficient balance to cover the amount wanting to stake.
+		InsufficientBalance,
+		/// Staker is attempting to stake an amount below the minimum amount.
+		InsufficientStakingAmount,
+		/// Staker is attempting to stake a zero amount.
+		ZeroAmountNotAllowed,
+		/// Origin has no Staking Account
+		NotAStakingAccount,
+		/// No staked value is available for withdrawal; either nothing is being unstaked,
+		/// or nothing has passed the thaw period.
+		NoUnstakedTokensAvailable,
+		/// Unstaking amount should be greater than zero.
+		UnstakedAmountIsZero,
+		/// Amount to unstake is greater than the amount staked.
+		AmountToUnstakeExceedsAmountStaked,
+		/// Attempting to unstake from a target that has not been staked to.
+		StakingAccountNotFound,
+		/// Attempting to get a staker / target relationship that does not exist.
+		StakerTargetRelationshipNotFound,
+		/// Attempting to get the target's capacity that does not exist.
+		TargetCapacityNotFound,
+		/// Staker reached the limit number for the allowed amount of unlocking chunks.
+		MaxUnlockingChunksExceeded,
+		/// Increase Capacity increase exceeds the total available Capacity for target.
+		IncreaseExceedsAvailable,
+		/// Attempting to set the epoch length to a value greater than the max epoch length.
+		MaxEpochLengthExceeded,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(current: T::BlockNumber) -> Weight {
+			Self::start_new_epoch_if_needed(current)
```

Comment:
It seems like weights `on_initialize` did not count 1 extra read that  `CurrentEpoch` has, since it is whitelisted. You should always assume that it's going to be read on each block.

Reply:
And whenever you are writing the epoch you should also add 1 extra write to your weights.

Reply:
Lets talk about this one.

Reply:
I think that would depend on how the benchmarks are written.  If they are set up so that it doesn't cross an epoch boundary I don't see how it would include that weight.

I had a PR that counted the read + write but closed it because it was so out of date rebasing was a nightmare.

Reply:
I don't think this is resolved, overhead benchmarks as far as I know does not account any whitelisted storages in extinsics. If we have **whitelisted** a storage we have to add  `1 DB_READ + 1 DB_WRITE `weight per block to account any access.

Reply:
Right. And that.